File: tests/csharp/tests-csharp-nunit-allure3-selenium/Helpers/ViewportHelper.cs
using Allure.NUnit.Attributes;
using Config;
using OpenQA.Selenium;
using OpenQA.Selenium.Chrome;

namespace Helpers;

public static class ViewportHelper
{
    [AllureStep("Reset viewport to default browser size")]
    public static void ResetViewport()
    {
        if (!WebDriverHolder.Has)
        {
            return;
        }

        var driver = WebDriverHolder.Get();
        if (!TryCdp(driver, "Emulation.clearDeviceMetricsOverride", new Dictionary<string, object>()))
        {
            driver.Manage().Window.Size = ParseBrowserSize(ConfigReader.TestConfig.BrowserSize);
        }
    }

    public static void SetViewport(int width, int height)
    {
        if (!WebDriverHolder.Has)
        {
            WebDrivers.StartBlank();
        }

        var driver = WebDriverHolder.Get();
        TryCdp(driver, "Emulation.clearDeviceMetricsOverride", new Dictionary<string, object>());
        var metrics = new Dictionary<string, object>
        {
            ["width"] = width,
            ["height"] = height,
            ["deviceScaleFactor"] = 1,
            ["mobile"] = false,
        };
        if (!TryCdp(driver, "Emulation.setDeviceMetricsOverride", metrics))
        {
            driver.Manage().Window.Size = new System.Drawing.Size(width, height);
        }
    }

    private static bool TryCdp(IWebDriver driver, string command, Dictionary<string, object> parameters)
    {
        if (driver is ChromeDriver chrome)
        {
            chrome.ExecuteCdpCommand(command, parameters);
            return true;
        }

        return false;
    }

    private static System.Drawing.Size ParseBrowserSize(string browserSize)
    {
        var parts = browserSize.Split('x');
        if (parts.Length != 2)
        {
            throw new InvalidOperationException("Invalid browserSize: " + browserSize);
        }

        return new System.Drawing.Size(int.Parse(parts[0].Trim()), int.Parse(parts[1].Trim()));
    }
}

File: tests/kotlin/tests-kotlin-gradle-junit5-allure3-selenium/src/test/java/helpers/ViewportHelper.java
package helpers;

import config.ConfigReader;
import io.qameta.allure.Step;
import org.openqa.selenium.Dimension;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.chromium.HasCdp;
import org.openqa.selenium.remote.Augmenter;

import java.util.Map;
import java.util.Optional;

public final class ViewportHelper {

    private ViewportHelper() {
    }

    @Step("Reset viewport to default browser size")
    public static void resetViewport() {
        if (!WebDriverHolder.has()) {
            return;
        }
        var driver = WebDriverHolder.get();
        resolveCdp(driver).ifPresentOrElse(
                cdp -> cdp.executeCdpCommand("Emulation.clearDeviceMetricsOverride", Map.of()),
                () -> driver.manage().window().setSize(parseBrowserSize(ConfigReader.testConfig.browserSize()))
        );
    }

    public static void setViewport(int width, int height) {
        if (!WebDriverHolder.has()) {
            WebDrivers.startBlank();
        }
        var driver = WebDriverHolder.get();
        resolveCdp(driver).ifPresent(
                cdp -> cdp.executeCdpCommand("Emulation.clearDeviceMetricsOverride", Map.of())
        );
        var metrics = Map.<String, Object>of(
                "width", width,
                "height", height,
                "deviceScaleFactor", 1,
                "mobile", false
        );
        resolveCdp(driver).ifPresentOrElse(
                cdp -> cdp.executeCdpCommand("Emulation.setDeviceMetricsOverride", metrics),
                () -> driver.manage().window().setSize(new Dimension(width, height))
        );
    }

    private static Optional<HasCdp> resolveCdp(WebDriver driver) {
        if (driver instanceof HasCdp hasCdp) {
            return Optional.of(hasCdp);
        }
        try {
            var augmented = new Augmenter().augment(driver);
            if (augmented instanceof HasCdp hasCdp) {
                return Optional.of(hasCdp);
            }
        } catch (RuntimeException ignored) {
            // Selenoid without CDP — fall back to window resize.
        }
        return Optional.empty();
    }

    private static Dimension parseBrowserSize(String browserSize) {
        var parts = browserSize.split("x");
        if (parts.length != 2) {
            throw new IllegalStateException("Invalid browserSize: " + browserSize);
        }
        return new Dimension(Integer.parseInt(parts[0].trim()), Integer.parseInt(parts[1].trim()));
    }
}
